README.md
ionic-service-deploy
====================

App deploy service for Ionic. Official docs: http://docs.ionic.io/services/deploy/


changelog.md
Changelog
=========

## 0.2.0

* ES6 Based build system
* Removed AngularJS dependencies
* Updated ionic-service-core dependency

## 0.1.2

* Updated core dependency version


## 0.1.2

* Settings now pulled from $ionicCoreSettings


## 0.1.1

* Added info method. Fixes driftyco/ionic-plugins-deploy#11


gulpfile.js
var gulp = require('gulp'),
  buildConfig = require('./build/config.js'),
  gutil = require('gulp-util'),
  concat = require('gulp-concat'),
  footer = require('gulp-footer'),
  header = require('gulp-header'),
  watch = require('gulp-watch'),
  browserify = require("browserify"),
  babelify = require("babelify"),
  eslint = require("gulp-eslint"),
  fs = require("fs")

gulp.task('build', ['lint'], function () {
  browserify({
    entries: buildConfig.jsFiles,
    debug: false,
    transform: [babelify]
  }).bundle()
  .on("error", function (err) { console.log("Error : " + err.message); })
  .pipe(fs.createWriteStream("ionic-deploy.js"));
});

gulp.task('lint', function () {
    return gulp.src(['src/**/*.js'])
        .pipe(eslint())
        .pipe(eslint.failOnError())
        .pipe(eslint.formatEach());
});

gulp.task('watch', ['build'], function() {
  gulp.watch(['src/**/*.js'], ['build']);
});

gulp.task('default', ['build']);


services.js
.factory('UpdateService', function($q) {
  return {
    check: function() {
      var deferred = $q.defer();

      IonicUpdate.check(function(result) {
        deferred.resolve(result === 'true');
      }, function(error) {
        deferred.reject(error);
      });

      return deferred.promise;
    },

    download: function() {
      var deferred = $q.defer();

      IonicUpdate.download(function(result) {
        if (result !== 'done') {
          deferred.notify(result);
        } else {
          deferred.resolve(result);
        }
      }, function(error) {
        deferred.reject(error);
      });

      return deferred.promise;
    },

    extract: function() {
      var deferred = $q.defer();

      IonicUpdate.extract(function(result) {
        if (result !== 'done') {
          deferred.notify(result);
        } else {
          deferred.resolve(result);
        }
      }, function(error) {
        deferred.reject(error);
      });

      return deferred.promise;
    },

    load: function() {
      IonicUpdate.redirect();
    },

    initialize: function(app_id) {
      IonicUpdate.initialize(app_id);
    }
  }
})


src/deploy.js
(function() {

  var Settings = new ionic.io.core.Settings();
  var DeferredPromise = ionic.io.util.DeferredPromise;

  var NO_PLUGIN = "IONIC_DEPLOY_MISSING_PLUGIN";
  var INITIAL_DELAY = 1 * 5 * 1000;
  var WATCH_INTERVAL = 1 * 60 * 1000;

  class Deploy {

    /**
     * Ionic Deploy
     *
     * This is the main interface that talks with the Ionic Deploy Plugin to facilitate
     * checking, downloading, and loading an update to your app.
     *
     * Base Usage:
     *
     *    var io = ionic.io.init();
     *    io.deploy.check().then(null, null, function(hasUpdate) {
     *      io.deploy.update();
     *    });
     *
     * @constructor
     */
    constructor() {
      var self = this;
      this._plugin = false;
      this._isReady = false;
      this._channelTag = 'production';
      this._emitter = ionic.io.core.main.events;
      console.log("Ionic Deploy: init");
      ionic.io.core.main.onReady(function() {
        self._isReady = true;
        self._emitter.emit('ionic_deploy:ready');
      });
    }

    /**
     * Fetch the Deploy Plugin
     *
     * If the plugin has not been set yet, attempt to fetch it, otherwise log
     * a message.
     *
     * @return {IonicDeploy} Returns the plugin or false
     */
    _getPlugin() {
      if (this._plugin) { return this._plugin; }
      if (typeof IonicDeploy === 'undefined') {
        console.log('Ionic Deploy: plugin is not installed or has not loaded. Have you run `ionic plugin add ionic-plugin-deploy` yet?');
        return false;
      }
      this._plugin = IonicDeploy;
      return IonicDeploy;
    }

    /**
     * Initialize the Deploy Plugin
     * @return {void}
     */
    initialize() {
      if (this._getPlugin()) {
        this._plugin.initialize(Settings.get('app_id'));
      }
    }

    /**
     * Check for updates
     *
     * @return {Promise} Will resolve with true if an update is available, false otherwise. A string or
     *   error will be passed to reject() in the event of a failure.
     */
    check() {
      var deferred = new DeferredPromise();

      if (this._getPlugin()) {
        this._plugin.check(Settings.get('app_id'), this._channelTag, function(result) {
          if (result && result === "true") {
            console.log('Ionic Deploy: an update is available');
            deferred.resolve(true);
          } else {
            console.log('Ionic Deploy: no updates available');
            deferred.resolve(false);
          }
        }, function(error) {
          console.log('Ionic Deploy: encountered an error while checking for updates');
          deferred.reject(error);
        });
      } else {
        deferred.reject(NO_PLUGIN);
      }

      return deferred.promise;
    }

    /**
     * Download and available update
     *
     * This should be used in conjunction with extract()
     * @return {Promise} The promise which will resolve with true/false or use
     *    notify to update the download progress.
     */
    download() {
      var deferred = new DeferredPromise();

      if (this._getPlugin()) {
        this._plugin.download(Settings.get('app_id'), function(result) {
          if (result !== 'true' && result !== 'false') {
            deferred.notify(result);
          } else {
            if (result === 'true') {
              console.log("Ionic Deploy: download complete");
            }
            deferred.resolve(result === 'true');
          }
        }, function(error) {
          deferred.reject(error);
        });
      } else {
        deferred.reject(NO_PLUGIN);
      }

      return deferred.promise;
    }


    /**
     * Extract the last downloaded update
     *
     * This should be called after a download() successfully resolves.
     * @return {Promise} The promise which will resolve with true/false or use
     *                   notify to update the extraction progress.
     */
    extract() {
      var deferred = new DeferredPromise();

      if (this._getPlugin()) {
        this._plugin.extract(Settings.get('app_id'), function(result) {
          if (result !== 'done') {
            deferred.notify(result);
          } else {
            console.log("Ionic Deploy: extraction complete");
            deferred.resolve(result);
          }
        }, function(error) {
          deferred.reject(error);
        });
      } else {
        deferred.reject(NO_PLUGIN);
      }

      return deferred.promise;
    }


    /**
     * Load the latest deployed version
     * This is only necessary to call if you have manually downloaded and extracted
     * an update and wish to reload the app with the latest deploy. The latest deploy
     * will automatically be loaded when the app is started.
     *
     * @return {void}
     */
    load() {
      if (this._getPlugin()) {
        this._plugin.redirect(Settings.get('app_id'));
      }
    }


    /**
     * Watch constantly checks for updates, and triggers an
     * event when one is ready.
     * @param {object} options Watch configuration options
     * @return {Promise} returns a promise that will get a notify() callback when an update is available
     */
    watch(options) {
      var deferred = new DeferredPromise();
      var opts = options || {};
      var self = this;

      if (typeof opts.initialDelay === 'undefined') { opts.initialDelay = INITIAL_DELAY; }
      if (typeof opts.interval === 'undefined') { opts.interval = WATCH_INTERVAL; }

      function checkForUpdates() {
        self.check().then(function(hasUpdate) {
          if (hasUpdate) { deferred.notify(hasUpdate); }
        }, function(err) {
          console.warn('Ionic Deploy: Unable to check for updates, ', err);
        });

        // Check our timeout to make sure it wasn't cleared while we were waiting
        // for a server response
        if (this._checkTimeout) {
          this._checkTimeout = setTimeout(checkForUpdates.bind(self), opts.interval);
        }
      }

      // Check after an initial short deplay
      this._checkTimeout = setTimeout(checkForUpdates.bind(self), opts.initialDelay);

      return deferred.promise;
    }

    /**
     * Stop automatically looking for updates
     * @return {void}
     */
    unwatch() {
      clearTimeout(this._checkTimeout);
      this._checkTimeout = null;
    }

    /**
     * Information about the current deploy
     *
     * @return {Promise} The resolver will be passed an object that has key/value
     *    pairs pertaining to the currently deployed update.
     */
    info() {
      var deferred = new DeferredPromise();

      if (this._getPlugin()) {
        this._plugin.info(Settings.get('app_id'), function(result) {
          deferred.resolve(result);
        }, function(err) {
          deferred.reject(err);
        });
      } else {
        deferred.reject(NO_PLUGIN);
      }

      return deferred.promise;
    }

    /**
     * Set the deploy channel that should be checked for updatse
     * See http://docs.ionic.io/docs/deploy-channels for more information
     *
     * @param {string} channelTag The channel tag to use
     * @return {void}
     */
    setChannel(channelTag) {
      this._channelTag = channelTag;
    }

    /**
     * Update app with the latest deploy
     *
     * @return {Promise} A promise result
     */
    update() {
      var deferred = new DeferredPromise();
      var self = this;

      if (this._getPlugin()) {
        // Check for updates
        self.check().then(function(result) {
          if (result === true) {
            // There are updates, download them
            var downloadProgress = 0;
            self.download().then(function(result) {
              if (!result) { deferred.reject("download error"); }
              self.extract().then(function(result) {
                if (!result) { deferred.reject("extraction error"); }
                self._plugin.redirect(Settings.get('app_id'));
              }, function(error) {
                deferred.reject(error);
              }, function(update) {
                var progress = downloadProgress + (update / 2);
                deferred.notify(progress);
              });
            }, function(error) {
              deferred.reject(error);
            }, function(update) {
              downloadProgress = (update / 2);
              deferred.notify(downloadProgress);
            });
          } else {
            deferred.resolve(false);
          }
        }, function(error) {
          deferred.reject(error);
        });
      } else {
        deferred.reject(NO_PLUGIN);
      }

      return deferred.promise;
    }

    /**
     * Fire a callback when deploy is ready. This will fire immediately if
     * deploy has already become available.
     *
     * @param {Function} callback Callback function to fire off
     * @return {void}
     */
    onReady(callback) {
      var self = this;
      if (this._isReady) {
        callback(self);
      } else {
        self._emitter.on('ionic_deploy:ready', function() {
          callback(self);
        });
      }
    }

  }

  ionic.io.register('deploy');
  ionic.io.deploy.DeployService = Deploy;

})();


ionicUpdate.js
angular.module('ionic.services.update', ['ionic.services.common'])

/**
 * @ngdoc service
 * @name $ionicUpdate
 * @module ionic.services.update
 * @description
 *
 * A simple way to push updates to your app.
 *
 * Initialize the service with your app id before calling other functions.
 * Then, use the check, download, extract and load functions to update and/or load
 * the updated version of your app.
 *
 * @usage
 * ```javascript
 * $ionicUpdate.initialize('8cdd99a2');
 * 
 * // Check for updates
 * $ionicUpdate.check().then(function(response) {
 *    // response will be true/false
 *    if (response) {
 *        // Download the updates
 *        $ionicUpdate.download().then(function() {
 *            // Extract the updates
 *            $ionicUpdate.extract().then(function() {
 *                // Load the updated version
 *                $ionicTrack.load();
 *            }, function(error) {
 *                // Error extracting
 *            }, function(progress) {
 *                // Do something with the zip extraction progress
 *                $scope.extraction_progress = progress;
 *            });
 *        }, function(error) {
 *            // Error downloading the updates
 *        }, function(progress) {
 *            // Do something with the download progress
 *            $scope.download_progress = progress;
 *        });
 *    }
 * } else {
 *    // No updates, load the most up to date version of the app
 *    $ionicUpdate.load();
 * }, function(error) {
 *    // Error checking for updates
 * })
 * ```
 */
.factory('$ionicUpdate', [
    '$q',
  function($q) {
    return {
      check: function() {
        var deferred = $q.defer();

        if (typeof IonicUpdate != "undefined") {
          IonicUpdate.check(function(result) {
            deferred.resolve(result === 'true');
          }, function(error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject("Plugin not loaded");
        }

        return deferred.promise;
      },

      download: function() {
        var deferred = $q.defer();

        if (typeof IonicUpdate != "undefined") {
          IonicUpdate.download(function(result) {
            if (result !== 'true' && result !== 'false') {
              deferred.notify(result);
            } else {
              deferred.resolve(result === 'true');
            }
          }, function(error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject("Plugin not loaded");
        }

        return deferred.promise;
      },

      extract: function() {
        var deferred = $q.defer();

        if (typeof IonicUpdate != "undefined") {
          IonicUpdate.extract(function(result) {
            if (result !== 'done') {
              deferred.notify(result);
            } else {
              deferred.resolve(result);
            }
          }, function(error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject("Plugin not loaded");
        }

        return deferred.promise;
      },

      load: function() {
        if (typeof IonicUpdate != "undefined") {
          IonicUpdate.redirect();
        }
      },

      initialize: function(app_id) {
        if (typeof IonicUpdate != "undefined") {
          IonicUpdate.initialize(app_id);
        }
      },

      update: function() {
        // This is an all-in-one function that's meant to do all of the update steps
        // in one shot.
        // NB: I think that the way to handle progress is to divide the provided progress result
        //     of each part by two (download and extract) and report that value.

        var deferred = $q.defer();

        if (typeof IonicUpdate != "undefined") {
          // Check for updates
          IonicUpdate.check(function(result) {
            if (result === 'true') {
              // There are updates, download them
              var progress = 0;
              IonicUpdate.download(function(result) {
                if (result !== 'true' && result !== 'false') {
                  // Download is only half of the reported progress
                  progress = progress + (result / 2);
                  deferred.notify(progress);
                } else {
                  // Download complete, now extract
                  IonicUpdate.extract(function(result) {
                    if (result !== 'done') {
                      // Extract is only half of the reported progress
                      progress = progress + (result / 2);
                      deferred.notify(progress);
                    } else {
                      // Extraction complete, now redirect
                      IonicUpdate.redirect();
                    }
                  }, function(error) {
                    // Error extracting updates
                    deferred.reject(error);
                  });
                }
              }, function(error) {
                // Error downloading updates
                deferred.reject(error);
              });
            } else {
              // There are no updates, redirect
              IonicUpdate.redirect();
            }
          }, function(error) {
            // Error checking for updates
            deferred.reject(error);
          });
        } else {
          deferred.reject("Plugin not loaded");
        }

        return deferred.promise;
      }
    }
}])


ionic-deploy.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
// Add Angular integrations if Angular is available
'use strict';

if (typeof angular === 'object' && angular.module) {
  angular.module('ionic.service.deploy', []).factory('$ionicDeploy', [function () {
    var io = ionic.io.init();
    return io.deploy;
  }]);
}

},{}],2:[function(require,module,exports){
"use strict";

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

(function () {

  var Settings = new ionic.io.core.Settings();
  var DeferredPromise = ionic.io.util.DeferredPromise;

  var NO_PLUGIN = "IONIC_DEPLOY_MISSING_PLUGIN";
  var INITIAL_DELAY = 1 * 5 * 1000;
  var WATCH_INTERVAL = 1 * 60 * 1000;

  var Deploy = (function () {

    /**
     * Ionic Deploy
     *
     * This is the main interface that talks with the Ionic Deploy Plugin to facilitate
     * checking, downloading, and loading an update to your app.
     *
     * Base Usage:
     *
     *    var io = ionic.io.init();
     *    io.deploy.check().then(null, null, function(hasUpdate) {
     *      io.deploy.update();
     *    });
     *
     * @constructor
     */

    function Deploy() {
      _classCallCheck(this, Deploy);

      var self = this;
      this._plugin = false;
      this._isReady = false;
      this._channelTag = 'production';
      this._emitter = ionic.io.core.main.events;
      console.log("Ionic Deploy: init");
      ionic.io.core.main.onReady(function () {
        self._isReady = true;
        self._emitter.emit('ionic_deploy:ready');
      });
    }

    /**
     * Fetch the Deploy Plugin
     *
     * If the plugin has not been set yet, attempt to fetch it, otherwise log
     * a message.
     *
     * @return {IonicDeploy} Returns the plugin or false
     */

    _createClass(Deploy, [{
      key: "_getPlugin",
      value: function _getPlugin() {
        if (this._plugin) {
          return this._plugin;
        }
        if (typeof IonicDeploy === 'undefined') {
          console.log('Ionic Deploy: plugin is not installed or has not loaded. Have you run `ionic plugin add ionic-plugin-deploy` yet?');
          return false;
        }
        this._plugin = IonicDeploy;
        return IonicDeploy;
      }

      /**
       * Initialize the Deploy Plugin
       * @return {void}
       */
    }, {
      key: "initialize",
      value: function initialize() {
        if (this._getPlugin()) {
          this._plugin.initialize(Settings.get('app_id'));
        }
      }

      /**
       * Check for updates
       *
       * @return {Promise} Will resolve with true if an update is available, false otherwise. A string or
       *   error will be passed to reject() in the event of a failure.
       */
    }, {
      key: "check",
      value: function check() {
        var deferred = new DeferredPromise();

        if (this._getPlugin()) {
          this._plugin.check(Settings.get('app_id'), this._channelTag, function (result) {
            if (result && result === "true") {
              console.log('Ionic Deploy: an update is available');
              deferred.resolve(true);
            } else {
              console.log('Ionic Deploy: no updates available');
              deferred.resolve(false);
            }
          }, function (error) {
            console.log('Ionic Deploy: encountered an error while checking for updates');
            deferred.reject(error);
          });
        } else {
          deferred.reject(NO_PLUGIN);
        }

        return deferred.promise;
      }

      /**
       * Download and available update
       *
       * This should be used in conjunction with extract()
       * @return {Promise} The promise which will resolve with true/false or use
       *    notify to update the download progress.
       */
    }, {
      key: "download",
      value: function download() {
        var deferred = new DeferredPromise();

        if (this._getPlugin()) {
          this._plugin.download(Settings.get('app_id'), function (result) {
            if (result !== 'true' && result !== 'false') {
              deferred.notify(result);
            } else {
              if (result === 'true') {
                console.log("Ionic Deploy: download complete");
              }
              deferred.resolve(result === 'true');
            }
          }, function (error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject(NO_PLUGIN);
        }

        return deferred.promise;
      }

      /**
       * Extract the last downloaded update
       *
       * This should be called after a download() successfully resolves.
       * @return {Promise} The promise which will resolve with true/false or use
       *                   notify to update the extraction progress.
       */
    }, {
      key: "extract",
      value: function extract() {
        var deferred = new DeferredPromise();

        if (this._getPlugin()) {
          this._plugin.extract(Settings.get('app_id'), function (result) {
            if (result !== 'done') {
              deferred.notify(result);
            } else {
              console.log("Ionic Deploy: extraction complete");
              deferred.resolve(result);
            }
          }, function (error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject(NO_PLUGIN);
        }

        return deferred.promise;
      }

      /**
       * Load the latest deployed version
       * This is only necessary to call if you have manually downloaded and extracted
       * an update and wish to reload the app with the latest deploy. The latest deploy
       * will automatically be loaded when the app is started.
       *
       * @return {void}
       */
    }, {
      key: "load",
      value: function load() {
        if (this._getPlugin()) {
          this._plugin.redirect(Settings.get('app_id'));
        }
      }

      /**
       * Watch constantly checks for updates, and triggers an
       * event when one is ready.
       * @param {object} options Watch configuration options
       * @return {Promise} returns a promise that will get a notify() callback when an update is available
       */
    }, {
      key: "watch",
      value: function watch(options) {
        var deferred = new DeferredPromise();
        var opts = options || {};
        var self = this;

        if (typeof opts.initialDelay === 'undefined') {
          opts.initialDelay = INITIAL_DELAY;
        }
        if (typeof opts.interval === 'undefined') {
          opts.interval = WATCH_INTERVAL;
        }

        function checkForUpdates() {
          self.check().then(function (hasUpdate) {
            if (hasUpdate) {
              deferred.notify(hasUpdate);
            }
          }, function (err) {
            console.warn('Ionic Deploy: Unable to check for updates, ', err);
          });

          // Check our timeout to make sure it wasn't cleared while we were waiting
          // for a server response
          if (this._checkTimeout) {
            this._checkTimeout = setTimeout(checkForUpdates.bind(self), opts.interval);
          }
        }

        // Check after an initial short deplay
        this._checkTimeout = setTimeout(checkForUpdates.bind(self), opts.initialDelay);

        return deferred.promise;
      }

      /**
       * Stop automatically looking for updates
       * @return {void}
       */
    }, {
      key: "unwatch",
      value: function unwatch() {
        clearTimeout(this._checkTimeout);
        this._checkTimeout = null;
      }

      /**
       * Information about the current deploy
       *
       * @return {Promise} The resolver will be passed an object that has key/value
       *    pairs pertaining to the currently deployed update.
       */
    }, {
      key: "info",
      value: function info() {
        var deferred = new DeferredPromise();

        if (this._getPlugin()) {
          this._plugin.info(Settings.get('app_id'), function (result) {
            deferred.resolve(result);
          }, function (err) {
            deferred.reject(err);
          });
        } else {
          deferred.reject(NO_PLUGIN);
        }

        return deferred.promise;
      }

      /**
       * Set the deploy channel that should be checked for updatse
       * See http://docs.ionic.io/docs/deploy-channels for more information
       *
       * @param {string} channelTag The channel tag to use
       * @return {void}
       */
    }, {
      key: "setChannel",
      value: function setChannel(channelTag) {
        this._channelTag = channelTag;
      }

      /**
       * Update app with the latest deploy
       *
       * @return {Promise} A promise result
       */
    }, {
      key: "update",
      value: function update() {
        var deferred = new DeferredPromise();
        var self = this;

        if (this._getPlugin()) {
          // Check for updates
          self.check().then(function (result) {
            if (result === true) {
              // There are updates, download them
              var downloadProgress = 0;
              self.download().then(function (result) {
                if (!result) {
                  deferred.reject("download error");
                }
                self.extract().then(function (result) {
                  if (!result) {
                    deferred.reject("extraction error");
                  }
                  self._plugin.redirect(Settings.get('app_id'));
                }, function (error) {
                  deferred.reject(error);
                }, function (update) {
                  var progress = downloadProgress + update / 2;
                  deferred.notify(progress);
                });
              }, function (error) {
                deferred.reject(error);
              }, function (update) {
                downloadProgress = update / 2;
                deferred.notify(downloadProgress);
              });
            } else {
              deferred.resolve(false);
            }
          }, function (error) {
            deferred.reject(error);
          });
        } else {
          deferred.reject(NO_PLUGIN);
        }

        return deferred.promise;
      }

      /**
       * Fire a callback when deploy is ready. This will fire immediately if
       * deploy has already become available.
       *
       * @param {Function} callback Callback function to fire off
       * @return {void}
       */
    }, {
      key: "onReady",
      value: function onReady(callback) {
        var self = this;
        if (this._isReady) {
          callback(self);
        } else {
          self._emitter.on('ionic_deploy:ready', function () {
            callback(self);
          });
        }
      }
    }]);

    return Deploy;
  })();

  ionic.io.register('deploy');
  ionic.io.deploy.DeployService = Deploy;
})();

},{}]},{},[2,1]);
